packages/api-server/src/price-oracle.ts

```diff
@@ -0,0 +1,281 @@
+      return this.pollTimeInterval;
+    }
+
+    return this.pollTimeInterval;
+  }
+
+  async price(): Promise<string | null> {
+    const price = await this.cacheStore.get(CKB_PRICE_CACHE_KEY);
+    return price;
+  }
+
+  // Return median gas_price of latest ${LATEST_MEDIAN_GAS_PRICE} transactions
+  async gasPrice(): Promise<bigint> {
+    // using cache
+    if (this.gasPriceCacheMilSec > 0) {
+      const cachedGasPrice = await this.cacheStore.get(GAS_PRICE_CACHE_KEY);
+      if (cachedGasPrice != null) {
+        return BigInt(cachedGasPrice);
+      }
+    }
+
+    let [medianGasPrice, minGasPrice] = await Promise.all([
+      this.query.getMedianGasPrice(),
+      this.minGasPrice(),
+    ]);
+    if (medianGasPrice < minGasPrice) {
+      medianGasPrice = minGasPrice;
+    }
+
+    // save cache
+    if (this.gasPriceCacheMilSec > 0) {
+      const medianGasPriceHex = "0x" + medianGasPrice.toString(16);
+      this.cacheStore.insert(
+        GAS_PRICE_CACHE_KEY,
+        medianGasPriceHex,
+        this.gasPriceCacheMilSec
+      );
+    }
+
+    return medianGasPrice;
+  }
+
+  async minGasPrice(): Promise<bigint> {
+    const ckbPrice = await this.price();
+    if (ckbPrice == null) {
+      // fallback to minimal
+      return MIN_GAS_PRICE_LOWER_LIMIT;
+    }
+    return Price.from(ckbPrice).toMinGasPrice();
+  }
+
+  async minFeeRate(): Promise<bigint> {
+    const ckbPrice = await this.price();
+    if (ckbPrice == null) {
+      // fallback to minimal
+      return MIN_GAS_PRICE_LOWER_LIMIT * FEE_RATE_MULTIPLIER;
+    }
+    return Price.from(ckbPrice).toMinFeeRate();
+  }
+
+  private async sendRequest(url: string) {
+    const validateStatus = function (status: number) {
+      return status === 200;
+    };
+    const options = {
+      validateStatus,
+      timeout: POLL_REQUEST_TIME_OUT,
+    };
+
+    let response = await axios.get(url, options).catch((error) => {
+      if (error.response) {
+        // The request was made and the server responded with wrong status code
+        throw new Error(
+          `[${CKBPriceOracle.name}] sendRequest: response failed, statusCode: ${error.response.status}, data: ${error.response.data}, header: ${error.response.headers}`
+        );
+      } else if (error.request) {
+        // The request was made but no response was received
+        // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
+        // http.ClientRequest in node.js
+        throw new Error(
+          `[${
+            CKBPriceOracle.name
+          }] sendRequest: request failed, ${JSON.stringify(error.toJSON())}`
+        );
+      } else {
+        // Something happened in setting up the request that triggered an Error
+        throw new Error(
+          `[${CKBPriceOracle.name}] sendRequest: setup request failed, ${error.message}`
+        );
+      }
+    });
+    return response.data;
+  }
+
+  private async pollPrice(): Promise<string | null> {
```

Comment:
For devnet, it's not necessary to query the price providers. Hence I suggest adding a switch configuration, default to `false`, to determine whether to query the price providers.

I want this configuration because the program prints errors every time I start it on my computer, making me careful.

Reply:
sounds fair, I will try later

Reply:
updated